packages/wrangler/src/api/types/index.ts

```diff
@@ -0,0 +1,71 @@
+import { writeFile } from "fs/promises";
+import path from "path";
+import { readConfig } from "../../config";
+import type { Config } from "../../config";
+
+export async function generateProjectTypes(options: {
+	config?: Config;
+	configFile?: string;
+	outFile?: string;
+	persistToFilesystem?: boolean;
+}): Promise<string> {
+	return generateTypes(getProjectTypes, options);
+}
+
+export async function generateRuntimeTypes(options: {
+	config?: Config;
+	configFile?: string;
+	outFile?: string;
+	persistToFilesystem?: boolean;
+}): Promise<string> {
+	return generateTypes(
+		async (config) =>
+			getRuntimeTypes({
+				compatibilityDate: config.compatibility_date,
+				compatibilityFlags: config.compatibility_flags.filter(
+					(flag) => !flag.includes("nodejs_compat")
+				),
+			}),
+		options
+	);
+}
+
+async function generateTypes(
+	getTypesFunction: (config: Config) => Promise<string>,
+	{
+		config,
+		configFile,
+		outFile,
+		persistToFilesystem = false,
+	}: {
+		config?: Config;
+		configFile?: string;
+		outFile?: string;
+		persistToFilesystem?: boolean;
+	}
+): Promise<string> {
+	if (!config && !configFile) {
+		throw new Error("Either config or configFile must be provided");
+	}
+	if (config && configFile) {
+		throw new Error("Only one of config or configFile should be provided");
+	}
+	if (outFile && persistToFilesystem === undefined) {
+		throw new Error(
+			"persistToFilesystem must be specified when outFile is provided"
+		);
+	}
```

Comment:
This combo have validation can be accomplished with union types for the options

Reply:
Or overloads perhaps.